### DISCUSSION TITLE
mysql data migration (customize table and column name)

### DISCUSSION THREAD

**Discussion by seventychi** [elsa-workflows/elsa-core#3431] [Q&A]
Hi,

Is there any way to customize table name and column name to match mysql naming convention (ex. WorkflowDefinitions => workflow_definition)

**Comment by seventychi** [comment]
I finally found an answer.
I create a class named ElsaDbContext, and inherited by ElsaContext, then override OnModelCreating, In OnModelCreating function rename all the table and columns like below.

`
protected override void OnModelCreating(ModelBuilder builder)
{
        #region WorkflowDefinition

        builder.Entity<WorkflowDefinition>().ToTable("workflow_definition");
        builder.Entity<WorkflowDefinition>().Property(p => p.Id).HasColumnName("id");
        builder.Entity<WorkflowDefinition>().Property(p => p.DefinitionId).HasColumnName("definition_id");
        builder.Entity<WorkflowDefinition>().Property(p => p.TenantId).HasColumnName("tenant_id");
        builder.Entity<WorkflowDefinition>().Property(p => p.Name).HasColumnName("name");
        builder.Entity<WorkflowDefinition>().Property(p => p.DisplayName).HasColumnName("display_name");
        builder.Entity<WorkflowDefinition>().Property(p => p.Description).HasColumnName("description");
        builder.Entity<WorkflowDefinition>().Property(p => p.Version).HasColumnName("version");
        builder.Entity<WorkflowDefinition>().Property(p => p.IsSingleton).HasColumnName("is_singleton");
        builder.Entity<WorkflowDefinition>().Property(p => p.PersistenceBehavior)
            .HasColumnName("persistence_behavior");
        builder.Entity<WorkflowDefinition>().Property(p => p.DeleteCompletedInstances)
            .HasColumnName("delete_completed_instances");
        builder.Entity<WorkflowDefinition>().Property(p => p.IsPublished).HasColumnName("is_published");
        builder.Entity<WorkflowDefinition>().Property(p => p.IsLatest).HasColumnName("is_latest");
        builder.Entity<WorkflowDefinition>().Property(p => p.Tag).HasColumnName("tag");
        builder.Entity<WorkflowDefinition>().Property(p => p.CreatedAt).HasColumnName("created_at");
        //builder.Entity<WorkflowDefinition>().Property("Data").HasColumnName("data");

        #endregion

        #region WorkflowInstance

        builder.Entity<WorkflowInstance>().ToTable("workflow_instance");
        builder.Entity<WorkflowInstance>().Property(p => p.Id).HasColumnName("id");
        builder.Entity<WorkflowInstance>().Property(p => p.DefinitionId).HasColumnName("definition_id");
        builder.Entity<WorkflowInstance>().Property(p => p.TenantId).HasColumnName("tenant_id");
        builder.Entity<WorkflowInstance>().Property(p => p.Version).HasColumnName("version");
        builder.Entity<WorkflowInstance>().Property(p => p.WorkflowStatus).HasColumnName("workflow_status");
        builder.Entity<WorkflowInstance>().Property(p => p.CorrelationId).HasColumnName("correlation_id");
        builder.Entity<WorkflowInstance>().Property(p => p.ContextType).HasColumnName("context_type");
        builder.Entity<WorkflowInstance>().Property(p => p.ContextId).HasColumnName("context_id");
        builder.Entity<WorkflowInstance>().Property(p => p.Name).HasColumnName("name");
        builder.Entity<WorkflowInstance>().Property(p => p.CreatedAt).HasColumnName("created_at");
        builder.Entity<WorkflowInstance>().Property(p => p.LastExecutedAt).HasColumnName("last_executed_at");
        builder.Entity<WorkflowInstance>().Property(p => p.FinishedAt).HasColumnName("finished_at");
        builder.Entity<WorkflowInstance>().Property(p => p.CancelledAt).HasColumnName("cancelled_at");
        builder.Entity<WorkflowInstance>().Property(p => p.FaultedAt).HasColumnName("faulted_at");
        builder.Entity<WorkflowInstance>().Property(p => p.LastExecutedActivityId)
            .HasColumnName("last_executed_activity_id");
        builder.Entity<WorkflowInstance>().Property(p => p.DefinitionVersionId)
            .HasColumnName("definition_version_id");
        //builder.Entity<WorkflowInstance>().Property("Data").HasColumnName("data");

        #endregion

        #region WorkflowExecutionLogRecord

        builder.Entity<WorkflowExecutionLogRecord>().ToTable("workflow_execution_log_record");
        builder.Entity<WorkflowExecutionLogRecord>().Property(p => p.Id).HasColumnName("id");
        builder.Entity<WorkflowExecutionLogRecord>().Property(p => p.TenantId).HasColumnName("tenant_id");
        builder.Entity<WorkflowExecutionLogRecord>().Property(p => p.WorkflowInstanceId)
            .HasColumnName("workflow_instance_id");
        builder.Entity<WorkflowExecutionLogRecord>().Property(p => p.ActivityId).HasColumnName("activity_id");
        builder.Entity<WorkflowExecutionLogRecord>().Property(p => p.ActivityType).HasColumnName("activity_type");
        builder.Entity<WorkflowExecutionLogRecord>().Property(p => p.Timestamp).HasColumnName("timestamp");
        builder.Entity<WorkflowExecutionLogRecord>().Property(p => p.EventName).HasColumnName("event_name");
        builder.Entity<WorkflowExecutionLogRecord>().Property(p => p.Message).HasColumnName("message");
        builder.Entity<WorkflowExecutionLogRecord>().Property(p => p.Source).HasColumnName("source");
        //builder.Entity<WorkflowExecutionLogRecord>().Property("Data").HasColumnName("data");

        #endregion

        #region Bookmark

        builder.Entity<Bookmark>().ToTable("bookmark");
        builder.Entity<Bookmark>().Property(p => p.Id).HasColumnName("id");
        builder.Entity<Bookmark>().Property(p => p.TenantId).HasColumnName("tenant_id");
        builder.Entity<Bookmark>().Property(p => p.Hash).HasColumnName("hash");
        builder.Entity<Bookmark>().Property(p => p.Model).HasColumnName("model");
        builder.Entity<Bookmark>().Property(p => p.ModelType).HasColumnName("model_type");
        builder.Entity<Bookmark>().Property(p => p.ActivityType).HasColumnName("activity_type");
        builder.Entity<Bookmark>().Property(p => p.ActivityId).HasColumnName("activity_id");
        builder.Entity<Bookmark>().Property(p => p.WorkflowInstanceId).HasColumnName("workflow_instance_id");
        builder.Entity<Bookmark>().Property(p => p.CorrelationId).HasColumnName("correlation_id");

        #endregion

        #region Trigger

        builder.Entity<Trigger>().ToTable("trigger");

        builder.Entity<Trigger>().Property(p => p.Id).HasColumnName("id");
        builder.Entity<Trigger>().Property(p => p.TenantId).HasColumnName("tenant_id");
        builder.Entity<Trigger>().Property(p => p.Hash).HasColumnName("hash");
        builder.Entity<Trigger>().Property(p => p.Model).HasColumnName("model");
        builder.Entity<Trigger>().Property(p => p.ModelType).HasColumnName("model_type");
        builder.Entity<Trigger>().Property(p => p.ActivityType).HasColumnName("activity_type");
        builder.Entity<Trigger>().Property(p => p.ActivityId).HasColumnName("activity_id");
        builder.Entity<Trigger>().Property(p => p.WorkflowDefinitionId).HasColumnName("workflow_definition_id");

        #endregion

        base.OnModelCreating(builder);
}
`

In Program.cs, 

`
services.AddElsa(elsa =>
        {
            elsa
                .AddConsoleActivities()
                .AddUserTaskActivities()
                .AddHttpActivities(config.GetSection("Server").Bind)
                .AddQuartzTemporalActivities()
                .AddWorkflowsFrom<T>()
                .UseEntityFrameworkPersistence<ElsaDbContext>(ef => ef.UseMySql(configuration.GetConnectionString("Default"))); // this line is important that change the default ElsaContext to our self-defined ElsaDbContext
        });
`

**Comment by johnwc** [comment]
MySQL does not have a naming convention for user databases, you use your own naming convention you set forth for your project. I highly discourage you from creating your own version of the MySQL provider. If there are any updates that the upstream introduces like new tables or new columns, your local version will not contain the new updates. You are creating a maintenance issue where there does not need to be. The tables and columns for this product is not meant for being read/written from by anything other than the workflow engine, and therefore should reside in its own database.
